File: metisfl/encryption/palisade/ckks_demo.cc

#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>

#include "ckks_scheme.h"

void generateRandomData(std::vector<double> &learner_Data, int rows, bool ceil_numbers = false) {

  double lower_bound = 0;
  double upper_bound = 100;

  std::uniform_real_distribution<double> unif(lower_bound, upper_bound);
  std::default_random_engine re;

  for (int i = 0; i < rows; i++) {

    if (ceil_numbers) {
      learner_Data.push_back(ceil(unif(re)));
    } else {
      learner_Data.push_back(unif(re));
    }

  }

}

int main() {

  uint32_t batchsize = 4096;
  uint32_t scalingfactorbits = 52;

  CKKS ckks(batchsize, scalingfactorbits);
  ckks.Print();

  // Whenever we change the `batchsize` and the `scalingfactorbits`
  // params we always need to invoke the GenCryptoContextAndKeys() function.
  auto tmp_dir = std::filesystem::temp_directory_path();
  CryptoParamsFiles crypto_params_files {
    tmp_dir / "cryptocontext.txt",
    tmp_dir / "key-public.txt",
    tmp_dir / "key-private.txt" };
  ckks.GenCryptoParams(crypto_params_files);
  PLOG(INFO) << crypto_params_files.crypto_context_file;
  ckks.LoadCryptoContextFromFile(crypto_params_files.crypto_context_file);
  ckks.LoadPublicKeyFromFile(crypto_params_files.public_key_file);
  ckks.LoadPrivateKeyFromFile(crypto_params_files.private_key_file);

  //generating random data for testing.
  vector<double> learner_data;
  generateRandomData(learner_data, 10, true);

  std::cout << "Learner Data: " << std::endl;
  std::cout << learner_data << std::endl << std::endl << std::endl;

  std::cout << "Encrypting" << std::endl;

  std::string enc_result = ckks.Encrypt(learner_data);
//  std::ofstream enc_result_fout("/tmp/metis/encrypted_random_numbers.out");
//  enc_result_fout << enc_result;
//  enc_result_fout.close();

  vector <std::string> learners_data;

  learners_data.push_back(enc_result);
  learners_data.push_back(enc_result);
  learners_data.push_back(enc_result);

  vector<float> scaling_factors;

  scaling_factors.push_back(0.5);
  scaling_factors.push_back(0.3);
  scaling_factors.push_back(0.5);

  std::cout << "Computing 0.5*L + 0.3*L + 0.5*L" << std::endl;

  std::string pwa_result =
      ckks.Aggregate(learners_data, scaling_factors);

  unsigned long int data_dimensions = learner_data.size();

  std::cout << "Decrypting" << std::endl;

  vector<double> pwa_res_pt = ckks.Decrypt(pwa_result, data_dimensions);

  std::cout << "Result:" << std::endl;

  std::cout << pwa_res_pt << std::endl << std::endl << std::endl << std::endl;

}

File: metisfl/encryption/demos/ckks_demo.py
import numpy as np
import os
import random

from fhe import CKKS
from metisfl.common.logger import MetisLogger


def encrypt(crypto_params_files, ckks_scheme, data):
    MetisLogger.info("Encrypting...")
    ckks_scheme.load_crypto_context_from_file(
        crypto_params_files["crypto_context_file"])
    ckks_scheme.load_public_key_from_file(
        crypto_params_files["public_key_file"])

    learners_data_encrypted = []
    for x in data:
        res_enc = ckks_scheme.encrypt(x)
        learners_data_encrypted.append(res_enc)
    MetisLogger.info("Encrypting is complete.")

    return learners_data_encrypted


def decrypt(crypto_params_files, ckks_scheme, data_enc, number_of_elems):
    # Make sure the input data is a list.
    if not isinstance(data_enc, list):
        data_enc = [data_enc]
    MetisLogger.info("Decrypting...")
    ckks_scheme.load_crypto_context_from_file(
        crypto_params_files["crypto_context_file"])
    ckks_scheme.load_private_key_from_file(
        crypto_params_files["private_key_file"])

    data_dec = []
    for x_enc in data_enc:
        res_dec = ckks_scheme.decrypt(x_enc, number_of_elems)
        # FIXME(hamzahsaleem): Why if the function call is not correct an encrypted message is generated?
        #  Just comment the line above and uncomment the one below to regenerate issue.
        # res_dec = ckks_scheme.decrypt(x_enc)
        data_dec.append(res_dec)
    MetisLogger.info("Decrypting is complete.")

    return data_dec


def pwa(crypto_params_files, ckks_scheme, data_enc, scaling_factors):
    MetisLogger.info("Computing Private Weighted Average...")
    ckks_scheme.load_crypto_context_from_file(
        crypto_params_files["crypto_context_file"])
    pwa_res = ckks_scheme.compute_weighted_average(
        data_enc, scaling_factors)
    MetisLogger.info("Private Weighted Average computation is complete.")
    return pwa_res


def test_ckks_api(batch_size, scaling_factor_bits, learners_data, scaling_factors, number_of_elems):
    MetisLogger.info("Generating crypto context and keys...")
    ckks_scheme = CKKS(batch_size, scaling_factor_bits)
    crypto_params_dir = os.path.join(
        os.getcwd(), "../resources/fheparams/cryptoparams")
    if not os.path.exists(crypto_params_dir):
        os.makedirs(crypto_params_dir)
    ckks_scheme.gen_crypto_context_and_keys(crypto_params_dir)
    crypto_params_files = ckks_scheme.get_crypto_params_files()
    MetisLogger.info("Crypto parameters files:")
    for param, file in crypto_params_files.items():
        MetisLogger.info("\t {}:{}".format(param, file))

    ckks_scheme = CKKS(batch_size, scaling_factor_bits)
    learners_data_enc = encrypt(
        crypto_params_files, ckks_scheme, learners_data)

    ckks_scheme = CKKS(batch_size, scaling_factor_bits)
    learners_data_dec = decrypt(
        crypto_params_files, ckks_scheme, learners_data_enc, number_of_elems)
    MetisLogger.info("Learners Data Decrypted: {}".format(learners_data_dec))

    ckks_scheme = CKKS(batch_size, scaling_factor_bits)
    pwa_enc = pwa(crypto_params_files, ckks_scheme,
                  learners_data_enc, scaling_factors)

    ckks_scheme = CKKS(batch_size, scaling_factor_bits)
    pwa_dec = decrypt(crypto_params_files, ckks_scheme,
                      pwa_enc, number_of_elems)
    MetisLogger.info("Aggregated (Decrypted) Result: {}".format(pwa_dec))


if __name__ == "__main__":
    """
    Through this demo we test the encryption, decryption and private weighted
    aggregation functions of the CKKS scheme. 

    To test each operation (encrypt, decrypt, pwa) in isolation and see which 
    crypto parameters are required to perform each operation, we create and 
    pass a separate CKKS scheme object at every function call. The corresponding 
    function loads the required crypto parameters through the given ckks scheme. 

    Specifically, for each operation we need the following crypto params: 
        encryption -> (crypto_context, public)
        decryption -> (crypto_context, private)
        pwa -> (crypto_context)     
    """

    # Define batch size and scaling factor bits of CKKS scheme.
    batch_size = 4096
    scaling_factor_bits = 52

    # Define demo example learners data and corresponding scaling factors.
    # In the first test case,
    number_of_learners = 2
    number_of_elems = 2 * batch_size
    learners_data = [[1 for _ in range(number_of_elems)]] * number_of_learners
    scaling_factors = [0.5] * number_of_learners
    # Just print the first 100 elements per learner.
    MetisLogger.info("Original learners data: {}".format(
        [x[:100] for x in learners_data]))
    MetisLogger.info("Original scaling factors: {}".format(
        np.array(scaling_factors[:number_of_learners])))
    test_ckks_api(batch_size, scaling_factor_bits, learners_data,
                  scaling_factors, number_of_elems)

    number_of_learners = 2
    number_of_elems = (2 * batch_size) + 1
    learners_data = [[2 for _ in range(number_of_elems)]] * number_of_learners
    scaling_factors = [0.5] * number_of_learners
    # Just print the first 100 elements per learner.
    MetisLogger.info("Original learners data: {}".format(
        [x[:100] for x in learners_data]))
    MetisLogger.info("Original scaling factors: {}".format(
        np.array(scaling_factors[:number_of_learners])))
    test_ckks_api(batch_size, scaling_factor_bits, learners_data,
                  scaling_factors, number_of_elems)
